### PR TITLE
Utilities for GWAS functions

### PR DESCRIPTION
**Pull request by eric-czech**
This adds a few utility functions necessary for https://github.com/pystatgen/sgkit/pull/66.

Summary of changes:

- Adds hypothesis as a test dependency
- Adds an r2 scoring function generalized to N-d with test results exported from scikit-learn (so it isn't a formal dependency)
- Adds a function for concatenating mixtures of 1D or 2D Dataset variables, which is a little more complicated than it sounds with xarray when the second dimension can differ (in name and length)

### CODE REVIEW

File: sgkit/tests/test_stats_utils.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+
+def test_concat_2d__raise_on_gte_3d():
+    ds = xr.Dataset(dict(x=(("dim0", "dim1", "dim2"), np.empty((1, 1, 1)))))
+    with pytest.raises(ValueError, match="All variables must have <= 2 dimensions"):
+        concat_2d(ds, dims=("dim0", "dim1"))
+
+
+@st.composite
+def sample_dataset(draw):
```
Comment (ravwojdyla) [inline sgkit/tests/test_stats_utils.py:93]: types missing

File: sgkit/tests/test_stats_utils.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    # Draw a mixture of 1D or 2D arrays with identical length along
+    # first axis (second axis may be absent or have size 0)
+    n = draw(st.integers(0, 10))
+    n_1d = draw(st.integers(1, 5))
+    n_2d = draw(st.integers(0, 5))
+    n_cols = draw(st.tuples(*[st.integers(0, 5)] * n_2d))
+    shapes = n_1d * [(n,)] + [(n, nc) for nc in n_cols]
```
Comment (ravwojdyla) [inline sgkit/tests/test_stats_utils.py:100]: There are some `numpy` specialised strategies in hypothesis - like `array_shapes`, the do not work here?

File: sgkit/tests/test_stats_utils.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    assert_array_shape(x, 10, 10)
+    with pytest.raises(AssertionError):
+        assert_array_shape(x, 9, 10)
+
+
+def _col_shape_sum(ds: Dataset) -> int:
+    return sum([1 if ds[v].ndim == 1 else ds[v].shape[1] for v in ds])
+
+
+def _rename_dim(ds: Dataset, prefix: str, name: str) -> Dataset:
```
Comment (ravwojdyla) [inline sgkit/tests/test_stats_utils.py:161]: Is this used anywhere?

File: sgkit/tests/test_stats_utils.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+
+def test_assert_array_shape():
+    x = da.zeros((10, 10))
+    assert_array_shape(x, 10, 10)
+    with pytest.raises(AssertionError):
+        assert_array_shape(x, 9, 10)
+
+
+def _col_shape_sum(ds: Dataset) -> int:
```
Comment (ravwojdyla) [inline sgkit/tests/test_stats_utils.py:157]: This seems to be used only once, why separate function?

File: sgkit/utils.py [hunk 5]
```diff
@@ -64,3 +64,55 @@ def encode_array(x: ArrayLike) -> Tuple[ArrayLike, List[Any]]:
     rank = np.empty_like(index)
     rank[index] = np.arange(len(index))
     return rank[inverse], names[index]
+
+
+def split_array_chunks(n: int, blocks: int) -> List[int]:
```
Comment (ravwojdyla) [inline sgkit/utils.py:69]: Would it make more sense to return a `Tuple`?

File: sgkit/utils.py [hunk 6]
```diff
@@ -64,3 +64,55 @@ def encode_array(x: ArrayLike) -> Tuple[ArrayLike, List[Any]]:
...
+    >>> split_array_chunks(7, 3)
+    [3, 2, 2]
+    >>> split_array_chunks(7, 1)
+    [7]
+    >>> split_array_chunks(7, 7)
+    [1, 1, 1, 1, 1, 1, 1]
+
+    Raises
+    ------
+    ValueError
```
Comment (ravwojdyla) [inline sgkit/utils.py:93]: Could this be stuck into a single `ValuerError`?

File: sgkit/tests/test_utils.py [hunk 7]
```diff
@@ -45,3 +45,39 @@ def check(x: ArrayLike, values: ArrayLike, names: List[Any]) -> None:
...
+        (7, 2, [4, 3]),
+        (7, 3, [3, 2, 2]),
+        (7, 7, [1] * 7),
+    ],
+)
+def test_split_array_chunks(n: int, blocks: int, expected_chunks: List[int]):
+    assert split_array_chunks(n, blocks) == expected_chunks
+
+
+def test_split_array_chunks__raise_on_blocks_gt_n():
```
Comment (ravwojdyla) [inline sgkit/tests/test_utils.py:68]: `split_array_chunks` could be tested with hypothesis as well - not that it must be, just saying :)

File: sgkit/stats/utils.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    Returns
+    -------
+    DataArray
+        Array with dimensions defined by `dims`.
+    """
+    arrs = []
+    for var in ds:
+        arr = ds[var]
+        if arr.dims[0] != dims[0]:
```
Comment (ravwojdyla) [inline sgkit/stats/utils.py:33]: where is this condition tested?

File: sgkit/utils.py [hunk 9]
```diff
@@ -64,3 +64,55 @@ def encode_array(x: ArrayLike) -> Tuple[ArrayLike, List[Any]]:
...
     rank[index] = np.arange(len(index))
     return rank[inverse], names[index]
+
+
+def split_array_chunks(n: int, blocks: int) -> List[int]:
+    """Compute chunk sizes for an array split into blocks.
+    This is similar to `numpy.split_array` except that it
+    will compute the sizes of the resulting splits rather
+    than explicitly partitioning an array.
+    Parameters
```
Comment (ravwojdyla) [inline sgkit/utils.py:74]: We should add a docstring linter. We can open a separate issue for this. But here we are missing a new line, also before `Examples`, and `Returns`.

File: sgkit/utils.py (hunk 5 above)
Comment (eric-czech) [reply sgkit/utils.py:69]: It could be Tuple[int, ...], why though?

File: sgkit/tests/test_stats_utils.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+from sgkit.stats.utils import (
+    assert_array_shape,
+    assert_block_shape,
+    assert_chunk_shape,
+    concat_2d,
+    r2_score,
+)
+
+
+def test_r2_score__batch_dims():
```
[the remark this answers is not in the document]
Comment (eric-czech) [reply sgkit/tests/test_stats_utils.py:20]: @tomwhite suggested it a while ago (though not as a hard rule), I have no strong opinion one way or the other except that it is kind of a nice visual break in the pytest output.

File: sgkit/utils.py (hunk 6 above)
Comment (eric-czech) [reply sgkit/utils.py:93]: In the code, docs or both?  I think I've seen a couple conventions for multiple conditions like that but I'm not sure what will work with Sphinx -- maybe something like this? https://github.com/dask/dask/blob/master/dask/array/core.py#L3320

```
    ValueError
        * If list depths are mismatched - for instance, ``[[a, b], c]`` is
          illegal, and should be spelt ``[[a, b], [c]]``
        * If lists are empty - for instance, ``[[a, b], []]``
```

File: sgkit/tests/test_utils.py (hunk 7 above)
Comment (eric-czech) [reply sgkit/tests/test_utils.py:68]: I think some of the boundary conditions could be but I don't see a good way to test the core of what the function does from random inputs (other than basically copying the code over and re-applying it in the test).  Let me know if you're thinking of something different.

File: sgkit/stats/utils.py (hunk 8 above)
Comment (eric-czech) [reply sgkit/stats/utils.py:33]: By this line: https://github.com/pystatgen/sgkit/blob/eac/stats_utils/sgkit/tests/test_stats_utils.py#L106

The first dimensions are assigned different labels there.

File: sgkit/utils.py (hunk 9 above)
Comment (eric-czech) [reply sgkit/utils.py:74]: That would be awesome, I think those keep getting introduced when I use the VSCode conflict resolution feature.  I'll fix that docstring at least.

Comment (eric-czech) [reply sgkit/utils.py:74]: Issue for it: https://github.com/pystatgen/sgkit/issues/70

File: sgkit/tests/test_stats_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/tests/test_stats_utils.py:20]: Oh, ok, I don't have anything against it, would be good to document it somewhere.

File: sgkit/utils.py (hunk 5 above)
Comment (ravwojdyla) [reply sgkit/utils.py:69]: Immutability, when both make sense, immutable data structure in my experience, tend to be less error prone and easier to work with.

File: sgkit/utils.py (hunk 6 above)
Comment (ravwojdyla) [reply sgkit/utils.py:93]: Just in the doc.

File: sgkit/tests/test_utils.py (hunk 7 above)
Comment (ravwojdyla) [reply sgkit/tests/test_utils.py:68]: @eric-czech 🤔 I think you could tests properties like:
 * size of the output equals the `blocks`
 * all elements are positive and monotonically increasing
 * sum equals to the `n`

File: sgkit/stats/association.py [hunk 11]
```diff
@@ -165,25 +129,36 @@ def gwas_linear_regression(
...
+        Dataset containing necessary dependent and independent variables.
     dosage : str
         Dosage variable name where "dosage" array can contain represent
         one of several possible quantities, e.g.:
         - Alternate allele counts
         - Recessive or dominant allele encodings
         - True dosages as computed from imputed or probabilistic variant calls
         - Any other custom encoding in a user-defined variable
-    covariates : Sequence[str]
+    covariates : Union[str, Sequence[str]]
```
[the remark this answers is not in the document]
Comment (eric-czech) [reply sgkit/stats/association.py:140]: https://github.com/pystatgen/sgkit/issues/72?  I don't mind it since it's part of the VSCode template but I do end up with a somewhat awkward mix of conventions when I'm describing arrays vs anything else.  Although "ArrayLike" has been good IMO, especially when it comes with dimension descriptions the way numpy/dask do it.

File: sgkit/stats/utils.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            arr = arr.rename({arr.dims[1]: dims[1]})
+        else:
+            # Add concatenation axis
+            arr = arr.expand_dims(dim=dims[1], axis=1)
+        arrs.append(arr)
+    return xr.concat(arrs, dim=dims[1])  # type: ignore[no-any-return,no-untyped-call]
+
+
+def r2_score(YP: ArrayLike, YT: ArrayLike) -> ArrayLike:
+    """R2 score calculator for batches of vector pairs
```
[the remark this answers is not in the document]
Comment (eric-czech) [reply sgkit/stats/utils.py]: https://github.com/pystatgen/sgkit/pull/69/commits/bda112f96279b909983e3b9c7e47234d6ea9a973

File: sgkit/stats/utils.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        All leading dimensions must be broadcastable to
+        the leading dimensions of `YP`.
+
+    Returns
+    -------
+    R2 : (...) ArrayLike
+        R2 scores array with shape equal to all leading
+        (i.e. batch) dimensions of the provided arrays.
+    """
+    assert YP.shape[-1] == YT.shape[-1]
```
[the remark this answers is not in the document]
Comment (eric-czech) [reply sgkit/stats/utils.py]: Whoops, missed that one.

https://github.com/pystatgen/sgkit/pull/69/commits/bda112f96279b909983e3b9c7e47234d6ea9a973

File: sgkit/tests/test_stats_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (eric-czech) [reply sgkit/tests/test_stats_utils.py:20]: Can you document it then if you think it's important?  I don't think it's something everybody should necessarily have to do.

File: sgkit/tests/test_stats_utils.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+def test_r2_score__sklearn_comparison(
+    predicted: List[Any], actual: List[Any], expected_r2: float
+):
+    yp, yt = np.array(predicted), np.array(actual)
+    assert r2_score(yp, yt) == expected_r2
+
+
+@given(st.integers(0, 25))
+@settings(max_examples=10)
+def test_concat_2d__values(n):
```
[the remark this answers is not in the document]
Comment (eric-czech) [reply sgkit/tests/test_stats_utils.py]: https://github.com/pystatgen/sgkit/pull/69/commits/bda112f96279b909983e3b9c7e47234d6ea9a973

File: sgkit/utils.py (hunk 5 above)
Comment (eric-czech) [reply sgkit/utils.py]: https://github.com/pystatgen/sgkit/pull/69/commits/bda112f96279b909983e3b9c7e47234d6ea9a973

File: sgkit/utils.py (hunk 6 above)
Comment (eric-czech) [reply sgkit/utils.py:95]: https://github.com/pystatgen/sgkit/pull/69/commits/bda112f96279b909983e3b9c7e47234d6ea9a973

File: sgkit/tests/test_utils.py (hunk 7 above)
Comment (eric-czech) [reply sgkit/tests/test_utils.py:80]: https://github.com/pystatgen/sgkit/pull/69/commits/bda112f96279b909983e3b9c7e47234d6ea9a973

File: sgkit/utils.py (hunk 9 above)
Comment (eric-czech) [reply sgkit/utils.py:75]: https://github.com/pystatgen/sgkit/pull/69/commits/bda112f96279b909983e3b9c7e47234d6ea9a973

**Comment by eric-czech** [comment]
@ravwojdyla do you know of a way to avoid doing this everywhere in tests:

```python
@given(st.integers(0, 25))  # type: ignore[misc]
@settings(max_examples=10)  # type: ignore[misc]
def test_something(n: int) -> None: ...
```

I get mypy errors w/o a return type (only for decorated functions) and I'm adding `ignore[misc]` for `Untyped decorator makes function "test_something" untyped`.  Is there some way to tell mypy to ignore type checking for any function wrapped by hypothesis decorators globally?

**Comment by ravwojdyla** [comment]
@eric-czech good question, maybe we can provide a stub in a `if TYPE_CHECKING` block in `conftest.py` or something like that. Would need to experiment with this.

**Comment by eric-czech** [comment]
> maybe we can provide a stub in a if TYPE_CHECKING block in conftest.py or something like that. Would need to experiment with this.

Hm ok sounds good.  Can you approve this @ravwojdyla if there's nothing else?  Or were you going to try that soon?

**Comment by ravwojdyla** [comment]
@eric-czech was just in the process of taking 2nd look. yea, let's not block on pytest config.

**Comment by ravwojdyla** [comment]
@eric-czech created an issue for that #74

**Comment by eric-czech** [comment]
Thanks @ravwojdyla.
